virttest/cpu.py

```diff
@@ -1258,3 +1261,111 @@ def get_qemu_best_cpu_model(params):
     # If no host cpu model can be found on qemu_cpu_models, choose the default
     return params.get("default_cpu_model", None)
+
+
+def get_qemu_best_cpu_info(params):
+    """
+    Get the best cpu model and features available for qemu
+    """
+    cpu_info = {}
+    cpu_model = get_qemu_best_cpu_model(params)
+    cpu_flags = ""
+    cpu_info["model"] = cpu_model
+    cpu_info["flags"] = cpu_flags
+    return cpu_info
+
+
+def convert_flags_from_str_to_dict(cpu_flags_str):
+    """
+    Convert cpu flags string to a dict
+    """
+    flags_dict = {}
+    if cpu_flags_str:
+        flags_list = cpu_flags_str.lstrip(",").split(",")
+        for item in flags_list:
+            if item.startswith("+"):
+                flags_dict[item[1:]] = "on"
+            elif item.startswith("-"):
+                flags_dict[item[1:]] = "off"
+            elif item.find("=") != -1:
+                item = item.split("=")
+                flags_dict[item[0]] = item[1]
+            else:
+                flags_dict[item] = None
+    return flags_dict
+
+
+def get_cpu_info_from_qemu_cli(name):
+    """
+    Get cpu info from qemu cmdline
+    """
+    cmd = "ps aux | grep qemu | grep -e '\s-name\s\+guest=%s\(\,\|\s\)'" % name
+    output = process.getoutput(cmd, shell=True)
+    cpu_line = re.findall("-cpu\s+(\S*)\s+", output)[0]
+    sep = cpu_line.index(",")
+    cpu_model = cpu_line[:sep]
+    cpu_flags = cpu_line[sep:]
+    cpu_info = {}
+    cpu_info["model"] = cpu_model
+    cpu_info["flags"] = cpu_flags
+    return cpu_info
+
+
+def get_cpu_info_from_virsh(params):
+    """
+    Try to get cpu model and features from virsh with 'host-model'
+    """
+    vm_arch = params["vm_arch_name"]
+    if vm_arch not in ['x86_64', 'i686']:
+        raise NotImplementedError("Arch '%s' is not supported" % vm_arch)
+    machine = params.get('machine_type')
+    name = uuid.uuid4().hex
+    try:
+        path.find_command("virsh")
+    except path.CmdNotFoundError:
+        logging.warning("Virsh executable not set or found on path")
+        return
+
+    xml = """
+            <domain type='kvm'>
+                <name>%s</name>
+                <memory>1</memory>
+                <os>
+                    <type arch='%s' machine='%s'>hvm</type>
+                </os>
+                <cpu mode='host-model'/>
+            </domain>
+          """ % (name, vm_arch, machine)
+    xml_file = os.path.join(data_dir.get_tmp_dir(), "temp_xml_for_cpu")
+    with open(xml_file, "w") as f:
+        f.write(xml)
+    try:
+        logging.info("Get cpu model and features from virsh")
+        virsh.define(xml_file)
+        virsh.start(name)
+    except Exception as err:
+        logging.error(err)
+        return
+    else:
+        cpu_info = get_cpu_info_from_qemu_cli(name)
+        return cpu_info
+    finally:
+        if virsh.is_alive(name):
+            virsh.destroy(name, ignore_status=True)
+        virsh.undefine(name, ignore_status=True)
```

Comment:
With the latest version,  use virsh start vm for every time, i think `ignore_status=True` is not suitable.

Reply:
Could you explain more? I don't get it.

Reply:
Define multiple vms with different name is ok, so after thinking about this more, i am ok with `ignore_status=True`.

One more questions, will these vms be cleaned at the end of test by force ?

Reply:
No. I'm not sure there is an alternative way to undefine. So far, I never hit undefine failure when domain exists.

Reply:
Yes, ignore_status = True is used to ignore the fails here. As my comment above, considering define multiple vms with different name is ok, so i am ok with ignore the fails in this step.  we are on the same page here.

My concern is about the failed situation, vm does not be destroyed even if no effect for the test.
But, after checked libvirt side,  i found `virsh destroy $domain` is destroy vm from host side with signal, and i think it is reliable to destroy vm, so no concern now, thanks!

Reply:
> Yes, ignore_status = True is used to ignore the fails here. As my comment above, considering define multiple vms with different name is ok, so i am ok with ignore the fails in this step. we are on the same page here.
> 
Actually I mean it's unlikely to fail undefine if define succeeds at beginning. So even vms have same names, there shouldn't have conflicts.

> My concern is about the failed situation, vm does not be destroyed even if no effect for the test.
> But, after checked libvirt side, i found `virsh destroy $domain` is destroy vm from host side with signal, and i think it is reliable to destroy vm, so no concern now, thanks!

Thanks for the check and review!